worker/worker.go

```diff
@@ -0,0 +1,239 @@
+	defer w.lmu.RUnlock()
+
+	if w.lifecycle == nil {
+		panic("TODO(worker): w.lifecycle is nil")
+	}
+
+	return w.lifecycle
+}
+
+// SetLifecycle blocks until a write lock is available on the guarding mutex,
+// and then updates it. The next item processed by the queue will be handled
+// using the new Lifecycle passed in as `l`.
+func (w *Worker) SetLifecycle(l Lifecycle) {
+	w.lmu.Lock()
+	defer w.lmu.Unlock()
+
+	w.lifecycle = l
+}
+
+// Close stops polling the queue immediately and waits for all tasks to complete
+// before stopping the heartbeat.
+func (w *Worker) Close() {
+	w.queue.Close()
+
+	w.closer <- struct{}{}
+	w.wg.Wait()
+
+	w.heart.Close()
+}
+
+// Halt stops the heartbeat and queue polling goroutines immediately and cancels
+// all tasks, marking them as FAILED before returning.
+func (w *Worker) Halt() {
+	w.closer <- struct{}{}
+
+	w.queue.Close()
+	w.heart.Close()
+
+	for i := 0; i < int(w.taskCount); i++ {
+		w.halter <- struct{}{}
+	}
+}
+
+// recv begins processing elements from the incoming queue, and delegating them
+// as Tasks along the wire. If an error occurs at the queue level, it will be
+// propagated up the chain to the Errs() channel on the Worker. Otherwise, if a
+// close or halt operation is given, recv will return immediately.
+//
+// This function runs in its own goroutine.
+func (w *Worker) recv() {
+	for {
+		select {
+		case payload := <-w.queue.In():
+			task := NewTask(payload, w.monitor(payload))
+			task.Begin()
+
+			w.tasks <- task
+		case err := <-w.queue.Errs():
+			w.errs <- err
+		case <-w.closer:
+			return
+		}
+	}
+}
+
+// monitor creates and returns a channel used to monitor when the Task is
+// marked as SUCCESSFUL or FAILED. Upon receiving a status update, it will
+// either call the Complete or Abandon function on the lifecycle. If a call to
+// Halt() is dispatched, the lifecycle will immediately abandon the current
+// task.
+func (w *Worker) monitor(payload []byte) (progress chan<- bool, closer <-chan struct{}) {
+	w.incrementTaskCount(1)
+
+	lifecycle := w.Lifecycle()
+	progress = make(chan bool)
+	closer = make(chan struct{})
+
+	go func(progress <-chan bool) {
+		defer w.decrementTaskCount()
+		defer close(progress)
+
+		var err error
+
+		select {
+		case success := <-progress:
+			if success {
+				err = lifecycle.Complete(payload)
+			} else {
+				err = lifecycle.Abandon(payload)
+			}
+		case <-w.halter:
+			closer <- struct{}{}
+			err = lifecycle.Abandon(payload)
+		}
+
+		if err != nil {
+			w.errs <- err
+		}
+
+	}(progress)
+
+	return
+}
+
+// incrementTaskCount adds the given delta to both the task count and the
+// WaitGroup in an atomic fashion.
+func (w *Worker) incrementTaskCount(delta int) {
+	w.wg.Add(delta)
+	atomic.AddUint64(&w.taskCount, uint64(delta))
+}
+
+// decrementTaskCount subtracts one from both the task count and the WaitGroup
+// in an atomic fashion.
+func (w *Worker) decrementTaskCount() {
+	w.wg.Done()
+	atomic.AddUint64(&w.taskCount, ^uint64(0))
+}
```

Comment:
hm. I really think using a map of tasks currently out would make life easier. Remember you'll also need to keep a mirrored copy of the worker's queue in Redis so that you can mark specific tasks as resolving or failing without scanning the entire list. You could then use a `sync.Cond` to lock and broadcast updates.